gapir/cc/stack.h

```diff
@@ -26,457 +26,172 @@
-      const void* pointer = stack->checkAndGetTopPointer("pop");
-      return const_cast<T*>(static_cast<const T*>(pointer));
-    }
-  };
-
   class Entry {
    public:
-    const void* valuePtr() const { return &mValue; }
-
-    template <typename T>
-    const T value() const {
-      static_assert(sizeof(mValue) >= sizeof(T),
-                    "T is too large to be used as value");
-      T t;
-      if (!getTo(&t)) {
-        GAPID_WARNING("Error: read stack value inappropriate type %s wanted %s",
-                      baseTypeName(mType),
-                      baseTypeName(TypeToBaseType<T>::type));
-        return T();
-      }
-      return t;
-    }
-
-    const BaseType& type() const { return mType; }
+    Entry()
+        : mDataSize(0), mDataType(typeid(void)), mIsVolatilePointer(false) {}
 
-    void set(bool b) {
-      mType = TypeToBaseType<bool>::type;
-      mValue.b = b;
-    }
-    void set(int8_t i8) {
-      mType = TypeToBaseType<int8_t>::type;
-      mValue.i8 = i8;
-    }
-    void set(int16_t i16) {
-      mType = TypeToBaseType<int16_t>::type;
-      mValue.i16 = i16;
-    }
-    void set(int32_t i32) {
-      mType = TypeToBaseType<int32_t>::type;
-      mValue.i32 = i32;
-    }
-    void set(int64_t i64) {
-      mType = TypeToBaseType<int64_t>::type;
-      mValue.i64 = i64;
-    }
-    void set(uint8_t u8) {
-      mType = TypeToBaseType<uint8_t>::type;
-      mValue.u8 = u8;
-    }
-    void set(uint16_t u16) {
-      mType = TypeToBaseType<uint16_t>::type;
-      mValue.u16 = u16;
-    }
-    void set(uint32_t u32) {
-      mType = TypeToBaseType<uint32_t>::type;
-      mValue.u32 = u32;
-    }
-    void set(uint64_t u64) {
-      mType = TypeToBaseType<uint64_t>::type;
-      mValue.u64 = u64;
-    }
-    void set(float f) {
-      mType = TypeToBaseType<float>::type;
-      mValue.f = f;
-    }
-    void set(double d) {
-      mType = TypeToBaseType<double>::type;
-      mValue.d = d;
+    template <class T>
+    static void assertCanStore() {
+      static_assert(
+          sizeof(T) <= maxEntrySize,
+          "You need to bump the value of gapir::Stack::Entry::maxEntrySize");
     }
 
-    template <typename T>
-    void set(T* p) {
-      mType = BaseType::AbsolutePointer;
-      mValue.p = const_cast<typename std::remove_const<T>::type*>(p);
-    }
+    template <class T>
+    void set(const T& val);
 
-    template <typename T>
-    void set(T val) {
-      static_assert(sizeof(T) == sizeof(mValue.u32),
-                    "Enum base type is not uint32_t");
-      mType = BaseType::Uint32;
-      mValue.u32 = uint32_t(val);
+    template <class T>
+    const T& get() {
+      assertCanStore<T>();
+#ifndef NDEBUG
+      // We only check the typeid for get() in debug builds. The rest of the
+      // time we just make sure the number of bytes is correct.
+      assert(sizeof(T) == mDataSize && mDataType == typeid(T));
+#else
+      assert(sizeof(T) == mDataSize);
+#endif
+      return *((T*)mData);
```

Comment:
This may segfault, depending on the data type.

Reply:
Changed to use std::aligned_storage.